### PR TITLE
Implement MappedTcpSocket::{map, new}

### PR DESCRIPTION
**Pull request by canndrew**
[Review on Reviewable](https://reviewable.io/reviews/maidsafe/nat_traversal/37)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @vinipsmaker

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/mapped_tcp_socket.rs [hunk 1]
```diff
@@ -36,18 +50,376 @@ pub struct MappedTcpSocket {
...
+                let mut stream = match mapping_socket.connect(&*simple_server) {
+                    Ok(stream) => stream,
+                    Err(e) => return Err(MappedTcpSocketMapWarning::MappingSocketConnect {
+                        addr: simple_server,
+                        err: e
+                    }),
+                };
+                let send_data = listener_message::REQUEST_MAGIC_CONSTANT;
+                // TODO(canndrew): What should we do if we get a partial write?
+                let _ = match stream.write(&send_data[..]) {
```
Comment (vinipsmaker) [inline src/mapped_tcp_socket.rs:360]: You can use [`write_all`](https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all).

**Comment by vinipsmaker** [comment]
It's failing to build on Rust stable: https://travis-ci.org/maidsafe/nat_traversal/jobs/111169007#L374

Updating `maidsafe_utilities::serialisation::SerialisationError` might solve the problem.

Rest of PR seems OK.

**Comment by canndrew** [comment]
Ah, right. It's depending on this: https://github.com/maidsafe/maidsafe_utilities/pull/26

**Comment by vinipsmaker** [comment]
> Ah, right. It's depending on this: maidsafe/maidsafe_utilities#26

Add a commit to update `Cargo.toml`.

**Comment by vinipsmaker** [comment]
It fails to build on Windows: https://ci.appveyor.com/project/MaidSafe-QA/nat-traversal/build/1.0.80/job/m5coigpfr0f867yo#L1509

**Comment by vinipsmaker** [comment]
Continuing at https://github.com/maidsafe/nat_traversal/pull/39

**Pull request closed without merging**